File: Step 10 Sliding Window & Two Pointer Combined Problems/Step 10.1 Medium Problems/Binary Subarrays With Sum/Binary Subarrays With Sum.java
class Solution {
    public int numSubarraysWithSum(int[] nums, int k) {
        HashMap<Integer, Integer> map = new HashMap<>();
        int sum = 0;
        int count = 0;
        for(int x: nums) {
            sum += x;
            if(sum==k)  count++;
            if(map.containsKey(sum-k)){
                count+= map.get(sum-k);
            }
            map.put(sum,map.getOrDefault(sum,0)+1);
        }
        return count;
    }
}

File: Step 10 Sliding Window & Two Pointer Combined Problems/Step 10.1 Medium Problems/Binary Subarrays With Sum/Binary Subarrays With Sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        return myHelper(nums, goal) - myHelper(nums, goal - 1);
    }
    int myHelper(vector<int>& nums, int goal) {
        if (goal < 0) return 0;
        int left = 0;
        int right = 0;
        int summ = 0;
        int count = 0;
        while (right < nums.size()) {
            summ += nums[right];
            while (summ > goal) summ -= nums[left++];
            count += (right - left + 1);
            right++;
        }
        return count;
    }
};